zookeeper-server/src/test/java/org/apache/zookeeper/server/auth/znode/groupacl/X509ZNodeGroupAclProviderTest.java

```diff
@@ -0,0 +1,174 @@
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class X509ZNodeGroupAclProviderTest extends ZKTestCase {
+  private static final Logger LOG = LoggerFactory.getLogger(X509ZNodeGroupAclProviderTest.class);
+  private static final String HOSTPORT = "127.0.0.1:" + PortAssignment.unique();
+  private static X509AuthTest.TestCertificate domainXCert;
+  private static X509AuthTest.TestCertificate superCert;
+  private static X509AuthTest.TestCertificate unknownCert;
+  private static final String CLIENT_CERT_ID_SAN_MATCH_TYPE = "6";
+  private static final String SCHEME = "x509";
+  private static ZooKeeperServer zks;
+  private ServerCnxnFactory serverCnxnFactory;
+  private ZooKeeper admin;
+  private static final String AUTH_PROVIDER_PROPERTY_NAME = "zookeeper.authProvider.x509";
+  private static final String CLIENT_URI_DOMAIN_MAPPING_ROOT_PATH = "/_CLIENT_URI_DOMAIN_MAPPING";
+  private static final String[] MAPPING_PATHS = {CLIENT_URI_DOMAIN_MAPPING_ROOT_PATH,
+      CLIENT_URI_DOMAIN_MAPPING_ROOT_PATH + "/CrossDomain",
+      CLIENT_URI_DOMAIN_MAPPING_ROOT_PATH + "/CrossDomain/SuperUser",
+      CLIENT_URI_DOMAIN_MAPPING_ROOT_PATH + "/DomainX",
+      CLIENT_URI_DOMAIN_MAPPING_ROOT_PATH + "/DomainX/DomainXUser",
+      CLIENT_URI_DOMAIN_MAPPING_ROOT_PATH + "/DomainY",
+      CLIENT_URI_DOMAIN_MAPPING_ROOT_PATH + "/DomainY/DomainYUser"};
+
+  @Before
+  public void setUp() throws Exception {
+    System.setProperty(X509ZNodeGroupAclProvider.ZOOKEEPER_ZNODEGROUPACL_SUPERUSER, "SuperUser");
+    System.setProperty("zookeeper.ssl.keyManager",
+        "org.apache.zookeeper.test.X509AuthTest.TestKeyManager");
+    System.setProperty("zookeeper.ssl.trustManager",
+        "org.apache.zookeeper.test.X509AuthTest.TestTrustManager");
+    System.setProperty(X509AuthenticationUtil.SSL_X509_CLIENT_CERT_ID_TYPE,
+        X509AuthenticationUtil.SUBJECT_ALTERNATIVE_NAME_SHORT);
+    System.setProperty(X509AuthenticationUtil.SSL_X509_CLIENT_CERT_ID_SAN_MATCH_TYPE,
+        CLIENT_CERT_ID_SAN_MATCH_TYPE);
+    System.setProperty(AUTH_PROVIDER_PROPERTY_NAME,
+        X509ZNodeGroupAclProvider.class.getCanonicalName());
+    System.setProperty(
+        ZNodeGroupAclProperties.ZNODE_GROUP_ACL_CONFIG_PREFIX + "clientUriDomainMappingRootPath",
+        CLIENT_URI_DOMAIN_MAPPING_ROOT_PATH);
+    LOG.info("Starting Zk...");
+    zks = new ZooKeeperServer(testBaseDir, testBaseDir, 3000);
+    final int PORT = Integer.parseInt(HOSTPORT.split(":")[1]);
+    serverCnxnFactory = ServerCnxnFactory.createFactory(PORT, -1);
+    serverCnxnFactory.startup(zks);
+    LOG.info("Waiting for server startup");
+    Assert.assertTrue("waiting for server being up ", ClientBase.waitForServerUp(HOSTPORT, 300000));
+    admin = ClientBase.createZKClient(HOSTPORT);
+    try {
+      ZKUtil.deleteRecursive(admin, CLIENT_URI_DOMAIN_MAPPING_ROOT_PATH);
+    } catch (Exception ignored) {
+    }
```

Comment:
Could probably throw the exception in the method signature?

Reply:
This exception has to be caught and not thrown. This is just a preventative measure to ensure the zk data tree is completely clean at the beginning of the test, otherwise there will be NODE_EXISTS exceptions. In the case that there's no residual nodes from last test (which is the normal case), deleteRecursive() will throw NO_NODE exceptions.

Reply:
Makes sense. Resolving.